### ISSUE TITLE
Mandatory casts for operations with ArrayXdual2nd

### ISSUE THREAD

**Issue by danilomendesdias** [autodiff/autodiff#97]
Dear @allanleal and other maintainers,

The following code doesn't compile
```
    auto x_2nd = ArrayXdual2nd(2);
    x_2nd << 1.0, 2.0;
    x_2nd = x_2nd + 1;
```
while 
```
    using dual2nd = HigherOrderDual<2>;

    auto x_2nd = ArrayXdual2nd(2);
    x_2nd << 1.0, 2.0;
    x_2nd = x_2nd + dual2nd(1);
```
works fine.

The same behavior doesn't happen with dual (1st order only), as the following also works
```
    auto x = ArrayXdual(2);
    x << 1.0, 2.0;
    x = x + 1;
```

I'm really impressed with this library and documentation. Very awesome work. Thank you for making it available.

**Comment by allanleal** [comment]
Hi @danilomendesdias - Thanks! 

What's the error you get?

**Comment by danilomendesdias** [comment]
The full output is huge, so I tried to extract just the errors. Let me know if you need everything:

```
In file included from ../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/Core:423:
../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/src/Core/functors/BinaryFunctors.h:34:77: error: no type named 'ReturnType' in 'Eigen::ScalarBinaryOpTraits<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double, Eigen::internal::scalar_sum_op<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double> >'
  typedef typename ScalarBinaryOpTraits<LhsScalar,RhsScalar,scalar_sum_op>::ReturnType result_type;


../../../examples/forward/playground.cpp:56:19: note: in instantiation of template class 'Eigen::CwiseBinaryOp<Eigen::internal::scalar_sum_op<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double>, const Eigen::Array<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, -1, 1, 0, -1, 1>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Array<double, -1, 1, 0, -1, 1> > >' requested here
    x_2nd = x_2nd + 1;
                  ^
In file included from ../../../examples/forward/playground.cpp:6:
In file included from ../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/Core:364:
../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/src/Core/util/Meta.h:312:39: error: no type named 'type' in 'std::result_of<Eigen::internal::scalar_sum_op<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double> (const autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> > &, const double &)>'
  typedef typename std::result_of<T>::type type1;

          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
In file included from ../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/Core:461:
../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/src/Core/CwiseBinaryOp.h:96:5: error: using declaration refers into 'Base::', which is not a base class of 'CwiseBinaryOp<Eigen::internal::scalar_sum_op<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double>, const Eigen::Array<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, -1, 1, 0, -1, 1>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Array<double, -1, 1, 0, -1, 1> > >'
    EIGEN_GENERIC_PUBLIC_INTERFACE(CwiseBinaryOp)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/src/Core/util/Macros.h:865:9: note: expanded from macro 'EIGEN_GENERIC_PUBLIC_INTERFACE'
  using Base::derived; \
        ^~~~~~

In file included from ../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/Core:461:
../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/src/Core/CwiseBinaryOp.h:96:5: error: using declaration refers into 'Base::', which is not a base class of 'CwiseBinaryOp<Eigen::internal::scalar_sum_op<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double>, const Eigen::Array<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, -1, 1, 0, -1, 1>, const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>, const Eigen::Array<double, -1, 1, 0, -1, 1> > >'
    EIGEN_GENERIC_PUBLIC_INTERFACE(CwiseBinaryOp)

../../../../../Work/miniconda/envs/myenv/include/eigen3/Eigen/src/Core/CwiseBinaryOp.h:107:7: error: static_assert failed due to requirement 'Eigen::internal::has_ReturnType<Eigen::ScalarBinaryOpTraits<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double, Eigen::internal::scalar_sum_op<autodiff::forward::Dual<autodiff::forward::Dual<double, double>, autodiff::forward::Dual<double, double> >, double> > >::value' "YOU_MIXED_DIFFERENT_NUMERIC_TYPES__YOU_NEED_TO_USE_THE_CAST_METHOD_OF_MATRIXBASE_TO_CAST_NUMERIC_TYPES_EXPLICITLY"
      EIGEN_CHECK_BINARY_COMPATIBILIY(BinaryOp,typename Lhs::Scalar,typename Rhs::Scalar);
```

**Comment by danilomendesdias** [comment]
Oh, sorry. just give me one second and I send you the minimal example.

**Comment by danilomendesdias** [comment]
That's the enough to get the errors:
```
#include <Eigen/Core>
using namespace Eigen;
#include <autodiff/forward.hpp>
#include <autodiff/forward/eigen.hpp>
using namespace autodiff;

using dual2nd = HigherOrderDual<2>;

int main()
{
    dual2nd T = 300;
    auto x = ArrayXdual2nd(2);
    x << 1.0, 2.0;

    x = x + 1;
}
```

**Comment by allanleal** [comment]
Hi Danilo, the issue has been fixed with PR #99 - thanks for reporting this!

**Comment by danilomendesdias** [comment]
Thank you very much, @allanleal. I'm enjoying it a lot.
